metacat-common-server/src/main/java/com/netflix/metacat/common/server/events/MetacatApplicationEventMulticaster.java

```diff
@@ -39,53 +43,79 @@
-     */
-    public void postAsync(final ApplicationEvent event) {
-        this.asyncEventMulticaster.multicastEvent(event);
+        this.registry = registry;
+        this.metacatProperties = metacatProperties;
     }
 
     /**
-     * Post event synchronously.
+     * Post event. Events will be handles synchronously or asynchronously based on the listener annotation.
      *
      * @param event event
      */
-    public void postSync(final ApplicationEvent event) {
+    public void post(final ApplicationEvent event) {
         super.multicastEvent(event);
+        asyncEventMulticasters.values().forEach(aem -> aem.multicastEvent(event));
     }
 
     @Override
     public void addApplicationListener(final ApplicationListener listener) {
-        super.addApplicationListener(listener);
-        asyncEventMulticaster.addApplicationListener(listener);
+        if (isAsyncListener(listener)) {
+            final Class<?> clazz = getListenerTargetClass(listener);
+            final String clazzName = clazz.getName();
+            if (!asyncEventMulticasters.containsKey(clazzName)) {
+                asyncEventMulticasters.put(clazzName, createApplicationEventMultiCaster(clazzName));
+            }
+            asyncEventMulticasters.get(clazzName).addApplicationListener(listener);
+        } else {
+            super.addApplicationListener(listener);
+        }
+    }
+
+    private boolean isAsyncListener(final ApplicationListener listener) {
+        return listener.getClass().isAnnotationPresent(AsyncListener.class)
+            || (listener instanceof MetacatApplicationListenerMethodAdapter
+                && ((MetacatApplicationListenerMethodAdapter) listener).getTargetClass()
+                .isAnnotationPresent(AsyncListener.class));
+    }
+
+    private Class<?> getListenerTargetClass(final ApplicationListener listener) {
+        return listener instanceof MetacatApplicationListenerMethodAdapter
+            ? ((MetacatApplicationListenerMethodAdapter) listener).getTargetClass() : listener.getClass();
+    }
+
+    private ApplicationEventMulticaster createApplicationEventMultiCaster(final String name) {
+        final SimpleApplicationEventMulticaster result = new SimpleApplicationEventMulticaster();
+        final ThreadPoolTaskExecutor executor = new ThreadPoolTaskExecutor();
+        executor.setCorePoolSize(metacatProperties.getEvent().getBus().getExecutor().getThread().getCount());
+        executor.initialize();
+        RegistryUtil.registerThreadPool(registry, "metacat.event.pool." + name,
+            executor.getThreadPoolExecutor());
+        result.setTaskExecutor(executor);
+        return result;
     }
 
     @Override
     public void removeApplicationListener(final ApplicationListener listener) {
         super.removeApplicationListener(listener);
-        asyncEventMulticaster.removeApplicationListener(listener);
+        asyncEventMulticasters.values().forEach(aem -> aem.removeApplicationListener(listener));
```

Comment:
Each listener can only belong to one asyncEventMulticaster. Should we get the key first and then remove from that instead of blindly calling remove it from all?